nexus_constructor/add_component_window.py

```diff
@@ -0,0 +1,258 @@
+                tooltip_on_accept="Units Valid",
+            )
+        )
+        self.unitsLineEdit.validator().isValid.connect(
+            self.ok_validator.set_units_valid
+        )
+
+        self.componentTypeComboBox.addItems(list(self.component_types.keys()))
+
+        # Validate the default value set by the UI
+        self.unitsLineEdit.validator().validate(self.unitsLineEdit.text(), 0)
+
+    def on_component_type_change(self):
+        self.webEngineView.setUrl(
+            QUrl(
+                f"http://download.nexusformat.org/sphinx/classes/base_classes/{self.componentTypeComboBox.currentText()}.html"
+            )
+        )
+        self.pixelOptionsBox.setVisible(
+            self.componentTypeComboBox.currentText() in PIXEL_COMPONENT_TYPES
+        )
+
+    def mesh_file_picker(self):
+        """
+        Opens the mesh file picker. Sets the file name line edit to the file path.
+        :return: None
+        """
+        filename = file_dialog(False, "Open Mesh", GEOMETRY_FILE_TYPES)
+        if filename:
+            self.fileLineEdit.setText(filename)
+            self.geometry_file_name = filename
+
+    def show_cylinder_fields(self):
+        self.geometryOptionsBox.setVisible(True)
+        self.geometryFileBox.setVisible(False)
+        self.cylinderOptionsBox.setVisible(True)
+
+    def show_no_geometry_fields(self):
+        self.geometryOptionsBox.setVisible(False)
+        if self.nameLineEdit.text():
+            self.buttonBox.setEnabled(True)
+
+    def show_mesh_fields(self):
+        self.geometryOptionsBox.setVisible(True)
+        self.geometryFileBox.setVisible(True)
+        self.cylinderOptionsBox.setVisible(False)
+
+    def generate_geometry_model(self):
+        """
+        Generates a geometry model depending on the type of geometry selected and the current values of the lineedits that apply to the particular geometry type.
+        :return: The generated model.
+        """
+        if self.CylinderRadioButton.isChecked():
+            geometry_model = CylinderModel()
+            geometry_model.set_unit(self.unitsLineEdit.text())
+            geometry_model.cylinder.height = self.cylinderHeightLineEdit.value()
+            geometry_model.cylinder.radius = self.cylinderRadiusLineEdit.value()
+            geometry_model.cylinder.axis_direction.setX(self.cylinderXLineEdit.value())
+            geometry_model.cylinder.axis_direction.setY(self.cylinderYLineEdit.value())
+            geometry_model.cylinder.axis_direction.setZ(self.cylinderZLineEdit.value())
+        if self.meshRadioButton.isChecked():
+            geometry_model = OFFModel()
+            geometry_model.set_units(self.unitsLineEdit.text())
+            geometry_model.set_file(self.geometry_file_name)
+        else:
+            geometry_model = NoShapeModel()
+        return geometry_model
+
+    def on_ok(self):
+        component_type = self.componentTypeComboBox.currentText()
```

Comment:
Does "type" here mean "Nx_class"? If so let's use the terminology `nx_class` to reduce ambiguity.